diff --git a/libs/locale/build.jam b/libs/locale/build.jam
--- a/libs/locale/build.jam
+++ b/libs/locale/build.jam
@@ -10,7 +10,8 @@ import feature ;

 # Features

 feature.feature boost.locale.iconv : on off : optional propagated ;
+feature.feature boost.locale.iconv.lib : libc libiconv : optional propagated ;
 feature.feature boost.locale.icu : on off :  optional propagated ;
 feature.feature boost.locale.posix : on off : optional propagated ;
 feature.feature boost.locale.std : on off : optional propagated ;
