diff --git a/docker/control/Dockerfile b/docker/control/Dockerfile
--- a/docker/control/Dockerfile
+++ b/docker/control/Dockerfile
@@ -1,4 +1,4 @@
-FROM jgoerzen/debian-base-standard:bookworm
+FROM debian:testing

 ENV container=docker
 STOPSIGNAL SIGRTMIN+3
@@ -6,10 +6,9 @@ STOPSIGNAL SIGRTMIN+3
 ENV LEIN_ROOT true

 # JDK21 only in Debian testing
 RUN echo "deb http://deb.debian.org/debian testing main" >> /etc/apt/sources.list
-ADD ./apt-preferences /etc/apt/preferences

 #
 # Jepsen dependencies
 #
 RUN apt-get -qy update && \
-    apt-get -qy install \
+    apt-get -qy -o Dpkg::Options::="--force-confnew" install \
         curl dos2unix emacs git gnuplot graphviz htop iputils-ping libjna-java openjdk-21-jdk-headless pssh screen vim wget
