diff --git a/pulsar-client-cpp/CMakeLists.txt b/pulsar-client-cpp/CMakeLists.txt
index 5cbbea6780b..5723d3e29d5 100644
--- a/pulsar-client-cpp/CMakeLists.txt
+++ b/pulsar-client-cpp/CMakeLists.txt
@@ -124,7 +124,10 @@ unset(OPENSSL_VERSION CACHE)
 if (LINK_STATIC)
     find_library(ZLIB_LIBRARIES REQUIRED NAMES libz.a z zlib)
     find_library(Protobuf_LITE_LIBRARIES NAMES libprotobuf-lite.a libprotobuf-lite)
-    find_library(CURL_LIBRARIES NAMES libcurl.a curl curl_a libcurl_a)
+    find_package(CURL REQUIRED)
+    if (${CMAKE_VERSION} VERSION_GREATER "3.12")
+        set(COMMON_LIBS ${COMMON_LIBS} CURL::libcurl)
+    endif ()
     find_library(LIB_ZSTD NAMES libzstd.a)
     find_library(LIB_SNAPPY NAMES libsnappy.a)
     message(STATUS "Protobuf_LITE_LIBRARIES: ${Protobuf_LITE_LIBRARIES}")
