diff --git a/CMakeLists.txt b/CMakeLists.txt
index b00f31d..17002f1 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1218,7 +1218,7 @@ if(NOT WIN32 OR ROCKSDB_INSTALL_ON_WINDOWS)

   if(ROCKSDB_BUILD_SHARED)
     install(
-      TARGETS ${ROCKSDB_SHARED_LIB}
+      TARGETS ${ROCKSDB_SHARED_LIB} OPTIONAL
       EXPORT RocksDBTargets
       COMPONENT runtime
       ARCHIVE DESTINATION "${CMAKE_INSTALL_LIBDIR}"
