--- a/include/libnuraft/attr_unused.hxx
+++ b/include/libnuraft/attr_unused.hxx
@@ -18,7 +18,7 @@
 #pragma once
 
 #ifndef ATTR_UNUSED
-#if defined(__linux__) || defined(__APPLE__)
+#if defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)
     #define ATTR_UNUSED __attribute__((unused))
 #elif defined(WIN32) || defined(_WIN32)
     #define ATTR_UNUSED
--- a/src/crc32.hxx
+++ b/src/crc32.hxx
@@ -22,7 +22,8 @@
 #define _JSAHN_CRC32_H
 
 #include <stdint.h>
-#if defined(__linux__) || defined(__APPLE__)
+#include <stddef.h>
+#if defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__)
     #include <unistd.h>
 #endif
 
